bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/service/ReadyServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.internal.service;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Consumer;
+
+import org.eclipse.smarthome.core.service.ReadyMarker;
+import org.eclipse.smarthome.core.service.ReadyMarkerFilter;
+import org.eclipse.smarthome.core.service.ReadyService;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of the {@link ReadyService} interface.
+ *
+ * @author Simon Kaufmann - initial contribution and API.
+ *
+ */
+@Component
+public class ReadyServiceImpl implements ReadyService {
+
+    private final Logger logger = LoggerFactory.getLogger(ReadyServiceImpl.class);
+    private static final ReadyMarkerFilter ANY = new ReadyMarkerFilter();
+
+    private final Set<ReadyMarker> markers = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+    private final Map<ReadyTracker, ReadyMarkerFilter> trackers = new HashMap<>();
+    private final ReentrantReadWriteLock rwlTrackers = new ReentrantReadWriteLock(true);
+
+    @Override
+    public void markReady(ReadyMarker readyMarker) {
+        rwlTrackers.readLock().lock();
+        try {
+            boolean isNew = markers.add(readyMarker);
+            if (isNew) {
+                notifyTrackers(readyMarker, tracker -> tracker.onReadyMarkerAdded(readyMarker));
+                logger.debug("Added ready marker {}", readyMarker);
+            }
+        } finally {
+            rwlTrackers.readLock().unlock();
+        }
+    }
+
+    @Override
+    public void unmarkReady(ReadyMarker readyMarker) {
+        rwlTrackers.readLock().lock();
+        try {
+            boolean isRemoved = markers.remove(readyMarker);
+            if (isRemoved) {
+                notifyTrackers(readyMarker, tracker -> tracker.onReadyMarkerRemoved(readyMarker));
+                logger.debug("Removed ready marker {}", readyMarker);
+            }
+        } finally {
+            rwlTrackers.readLock().unlock();
+        }
+    }
+
+    private void notifyTrackers(ReadyMarker readyMarker, Consumer<ReadyTracker> action) {
+        trackers.entrySet().stream().filter(entry -> {
```

Comment:
Not a change request but a suggestion: 

```java
trackers.entrySet().stream().filter(entry -> entry.getValue().apply(readyMarker)).map(entry -> entry.getKey())
.forEach(action);
````

Reply:
or

```java
trackers.entrySet().stream().filter(entry -> entry.getValue().apply(readyMarker)).map(Entry::getKey)
.forEach(action);
```

Reply:
I've been using the block-notation on purpose, to force the auto-formatter to insert line-breaks in places where the human brain can follow. I couldn't figure out how to make JDT do do something similar automatically. Do you by any change have an idea?

Reply:
Ah now I understand. What I usually do to achieve this is that I manually make changes to the stream notations and turn of autoformatter for certain section.

```java
// @formatter:off
trackers.entrySet().stream()
                    .filter(entry -> entry.getValue().apply(readyMarker))
                    .map(Entry::getKey)
                    .forEach(action);
// @formatter:on
```